omnipath/_misc/dtypes.py

```diff
@@ -0,0 +1,175 @@
+from typing import Union, Iterable
+
+import pandas as pd
+import numpy as np
+
+
+TRUE = frozenset(('true', 't', 'yes', 'y'))
+FALSE = frozenset(('false', 'f', 'no', 'n'))
+BOOL = frozenset().union(TRUE, FALSE)
+NA = frozenset((
+    'na', 'NA', 'NaN',
+    'none', 'None', None,
+    pd.NA, pd.NaT, np.NAN, np.nan
+))
+INT = ('int64', 'uint64')
+ALL = INT + ('float64', 'string')
+
+
+def auto_dtype(
+    data: Union[pd.DataFrame, pd.Series, Iterable],
+    convert: bool = True,
+    categories: bool = True,
+    **kwargs,
+) -> Union[pd.DataFrame, pd.Series, str, dict]:
+    """
+    Automatically guesses and optionally converts data types of a dataframe,
+    series or other iterable.
+
+    Parameters
+    ----------
+    data
+        A dataframe or an array like object such as :class:`pandas.Series`,
+        :class:`numpy.ndarray` or list.
+    convert
+        Convert the data types to the best fitting types.
+    categories
+        Use the `category` data type for string variables with a small
+        number of values compared to their size.
+    **kwargs
+        For dataframes, manually set the desired data type of certain
+        variables.
+
+    Returns
+    -------
+    :class:`pandas.DataFrame` or :class:`pandas.Series` or str or list
+        If :arg:`convert` is `True`: a dataframe or series with its data
+        type(s) converted.
+        If :arg:`convert` is `False`: Name of a data type for a single Series
+        or iterable, or a dict of data type names by column labels for a
+        dataframe.
+    """
+
+    method = (
+        _auto_dtype_df
+            if isinstance(data, pd.DataFrame) else
+        _auto_dtype_series
+    )
+
+    return method(data, convert = convert, categories = categories, **kwargs)
+
+
+def _auto_dtype_df(
+    data: pd.DataFrame,
+    convert: bool = True,
```

Comment:
Am wondering why `convert: bool = True` even exists: if `convert = False` and the column is in `kwargs`, you return the `dtype` you would convert it to. If `convert = False` and the column is not in `kwargs`, you get the would-be `dtype` from the current series.
Not sure why it's important to keep this option, I think it doesn't add anything, but maybe I'm missing something. However if not, would remove this behavior.